nonce/nonce.go

```diff
@@ -259,18 +259,19 @@ func (ns *NonceService) Valid(nonce string) bool {
 		return false
 	}
 
-	ns.mu.Lock()
-	defer ns.mu.Unlock()
-	if c > ns.latest {
+	if c > ns.latest.Load() {
 		ns.nonceRedeems.WithLabelValues("invalid", "too high").Inc()
 		return false
 	}
 
-	if c <= ns.earliest {
+	if c <= ns.earliest.Load() {
 		ns.nonceRedeems.WithLabelValues("invalid", "too low").Inc()
 		return false
 	}
 
+	ns.mu.Lock()
+	defer ns.mu.Unlock()
```

Comment:
It totally makes sense that we don't need to lock in order to access ns.latest and ns.earliest, since those are now atomics. However, moving this lock to after those accesses means that it is possible for ns.earliest to be updated *in between* those checks. (It's also possible for ns.latest to be updated in between the checks, but that doesn't matter because we'd have already checked ns.latest at that point.)

I don't think this is necessarily a problem. But it is a change. Now the following sequence of events could happen:
- ns.earliest is 10; ns.latest is 20; ns.used is full
- request A arrives carrying nonce 11
- request B arrives carrying nonce 15
- request A passes the "not greater than latest" check
- request B passes the "not greater than latest" check
- request B passes the "no less than earliest" check, grabs the lock
- request B updates earliest to e.g. 12 (because ns.used was full)
- request A fails the "not less than earliest" check

Previously, request A would have grabbed the lock before passing the "not greater than latest" check, and both requests would have succeeded. Of course, in the old world it was totally possible for request B to have beaten request A to the punch, and caused request A to fail anyway. So I don't think this is a meaningful problem.

It's just rather surprising that the the goalposts can move in the middle of the checks. There's no data race here thanks to the atomics, but there is a logic race.

Reply:
Actually maybe that is bad:

We’re looking at N+1 and it passes the earliest check
Earliest moves forward, say from N to n+2. 
Acquire lock.
Now we put N+1 in the heap, which is below earliest, so on the next pop it’ll go backwards!

Reply:
Ah, a race not between passing the latest check and passing the earliest check, but between passing the earliest check and updating the value of earliest. Yep, agreed, that means the lock needs to come before the ns.earliest check at least.

From a "does this do what it intuitively says it does" perspective, I'd still _kinda_ like the lock to come before the ns.latest check, and to have nonce redemption still require a lock before incrementing ns.latest. I still don't think that has a meaningful impact on correctness, just on matching expectations.

Reply:
I moved earliest inside the lock before reading this comment.  I don't think latest needs to be inside the lock.

Unlike earliest, there's no additional datastructures we're comparing it with.  We're either incrementing when generating a nonce, which has no benifits of locking, or in Valid, we load then compare, which also doesn't matter if it increments from new generated nonces in the rest of the Valid function.